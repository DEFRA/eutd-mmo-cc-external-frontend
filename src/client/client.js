// Startup point for the client side application
import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import {Router, withRouter} from 'react-router-dom';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { renderRoutes } from 'react-router-config';
import axios from 'axios';
import Routes from './Routes';
import reducers from './reducers';
import { createBrowserHistory } from 'history';

const composeEnhancers = process.env.NODE_ENV === 'development' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__  ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose;
  

const paramSerializerFn = params => {
  let output = '';
  Object.keys(params).forEach((k) => {
    output+= `${k}=${params[k]}&`;
  });
  return encodeURI(output);
};

const axiosResponseInterceptorSuccessFn = response => response;
const axiosResponseInterceptorErrorFn = error => {
  if (error.response.status === 599) {
    // Let's check if this is because we're logged out
    const empPromise = new Promise(() => {});
    const searchParam = error.request.responseURL.slice(error.request.responseURL.search(/[?]/g));
    let searchParams = new URLSearchParams(searchParam);
    if(searchParams.get('notLoggedInErr') === 'yes') {
      window.location = '/logout?backToPath=/server-logout';
      return empPromise;
    }
    // It's always a timeout...
    window.location = '/timed-out';
    return new Promise(() => {});
  } else {
    return new Promise((_,rej) => rej(error));
  }
};

const axiosInstanceForOrchService = axios.create({
  baseURL: '/orchestration/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForOrchService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForRefService = axios.create({
  baseURL: '/reference/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForRefService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForDynamixService = axios.create({
  baseURL: '/dynamix/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForDynamixService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const store = createStore(
  reducers,
  window.INITIAL_STATE,
  composeEnhancers(applyMiddleware(
    // The difference between orch/ref service api is not significant as they both go through proxy
    // However it is significant in the server side store. To stick with same action creators we're
    // passing down same axios instance to both orchestration and reference service api .
    thunk.withExtraArgument( {
      orchestrationApi: axiosInstanceForOrchService,
      referenceServiceApi: axiosInstanceForRefService,
      dynamixApi: axiosInstanceForDynamixService } )
  ))
);

const ScrollToTop = withRouter( class ScrollToTop extends React.Component {
  componentDidMount() {
  }

  render() {
    return this.props.children;
  }
});

const history = createBrowserHistory();


ReactDOM.hydrate(
  <Provider store={store}>
    <Router history={history}>
      <ScrollToTop>
        <div>{renderRoutes(Routes)}</div>
      </ScrollToTop>
    </Router>
  </Provider>,
  document.querySelector('#root')
);
