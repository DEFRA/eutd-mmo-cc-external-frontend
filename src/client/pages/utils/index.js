import React, {Fragment} from 'react';

export * from './errorUtils';


export function formatAddress( ...adresssLines) {
  return adresssLines.filter( a => a ).map( (a, index) => <Fragment key={index}>{a}<br/></Fragment> );
}

export function constructAddress(addressArray) {
  let address = '';
  if (addressArray.length === 4) {
    if (addressArray[0]) {
      address += addressArray[0];
    }
    if (addressArray[0] && addressArray[1]) {
      address += ' ';
    }
    if (addressArray[1]) {
      address += addressArray[1];
    }
    if (addressArray[2]) {
      address += ', ' + addressArray[2];
    }
    if (addressArray[3]) {
      address += '. ' + addressArray[3];
    }
  }
  return address;
}

export const hasAdminOverride = (exportPayload) => {
  const isLandingOverridden = (landing) =>
    landing &&
    landing.model &&
    landing.model.vessel &&
    landing.model.vessel.vesselOverriddenByAdmin;

  return exportPayload &&
    exportPayload.items &&
    exportPayload.items.some(
      (item) =>
        item.landings &&
        item.landings.some((landing) => isLandingOverridden(landing))
    )
    ? true
    : false;
};