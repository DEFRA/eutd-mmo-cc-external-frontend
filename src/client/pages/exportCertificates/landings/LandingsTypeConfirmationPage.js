import React from 'react';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import _ from 'lodash';

import {
  Main,
  BackLink,
  Header,
  GridRow,
  GridCol,
  MultiChoice,
  WarningText,
} from 'govuk-react';

import { onHandleErrorClick, scrollToErrorIsland } from '../../utils';
import {
  clearLandingsType,
  confirmChangeLandingsType,
  onLoadComponentRedirect,
  clearChangedLandingsType,
 } from '../../../actions/landingsType.actions';
import { dispatchApiCallFailed, dispatchClearErrors } from '../../../actions';
import { withTranslation } from 'react-i18next';
import PageTitle from '../../../components/PageTitle';
import Form from '../../../components/elements/Form';
import HelpLink from '../../../components/HelpLink';
import SecondaryButton from '../../../components/elements/SecondaryButton';
import ContinueButton from '../../../components/elements/ContinueButton';
import ErrorIsland from '../../../components/elements/ErrorIsland';
import SelectRadio from '../../../components/elements/SelectRadio';
import PageTemplateWrapper from '../../../components/PageTemplateWrapper';

class LandingsTypesConfirmationPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      confirmLandingsType: ''
    };
    this.unauthorised.bind(this);
  }

  onChange = e => {
    this.setState({ confirmLandingsType: e.target.value });
  };

  onClear = (e) => {
    e.preventDefault();
    const { history, route, match } = this.props;

    this.props.dispatchClearErrors();
    history.push(route.previousUri.replace(':documentNumber', match.params.documentNumber));
  }

  displayDashboardOptions() {
    const { history, route } = this.props;
    history.push(route.dashboardUri);
  }

  componentDidMount() {
    const { landingsType = {}, changedLandingsType } = this.props;

    if (landingsType.unauthorised) {
      this.unauthorised();
    }

    if (changedLandingsType === '') {
      this.displayDashboardOptions();
    }
  }

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  onSubmit = async e => {
    e.preventDefault();

    try {
      const { route, match } = this.props;
      const documentNumber = match.params['documentNumber'];
      const { changedLandingsType } = this.props;
      const data = {
        landingsEntryOption: changedLandingsType,
        landingsEntryConfirmation: this.state.confirmLandingsType
      };
      await this.props.confirmChangeLandingsType(data, route.journey, route.path, documentNumber);
      if (this.state.confirmLandingsType === 'Yes') {
        this.props.history.push(route.nextUri.replace(':documentNumber', documentNumber));
      } else {
        this.props.history.push(route.previousUri.replace(':documentNumber', documentNumber));
      }
    } catch (error) {
      scrollToErrorIsland();
    }
  };

  componentWillUnmount() {
    this.props.clearLandingsType();
    this.props.clearChangedLandingsType();
  }

  render() {
    const { confirmLandingsChangeError, errors } = this.props.errors;
    const { route, match, history, t } = this.props;
    const { confirmLandingsType } = this.state;
    const { path, previousUri, title } = route;
    const heading = t('ccLandingsTypeConfirmationHeaderText');
    const currentUri = path.replace(
      ':documentNumber',
      match.params.documentNumber
    );
    return (
      <Main>
        {errors && (
          <ErrorIsland
            errors={errors.map((err) => ({
              message: t(err.text),
              key: err.targetName,
            }))}
            onHandleErrorClick={onHandleErrorClick}
          />
        )}

        <PageTitle
          title={`${
            !_.isEmpty(errors) ? `${t('commonPageTitleErrorText')} ` : ''
          }${heading} - ${t(title)}`}
        />
        <GridRow>
          <BackLink
            onClick={(e) => {
              e.preventDefault();
              this.props.dispatchClearErrors();
              history.push(previousUri.replace(':documentNumber', match.params.documentNumber));
            }}
            href={previousUri.replace(':documentNumber', match.params.documentNumber)}
          >{t('commonBackLinkBackButtonLabel')}</BackLink>
        </GridRow>

        <Form
          action="/orchestration/api/v1/confirm-change-landings-type"
          currentUrl={currentUri}
          onSubmit={this.onSubmit}
        >
            <GridCol columnFull>
              <GridRow>
                <Header level="1">
                  <legend>{heading}</legend>
                </Header>
              </GridRow>
              <GridRow>
                <WarningText className="warning-message">
                  {t('ccLandingsTypeConfirmationWarningText')}
                </WarningText>
              </GridRow>
              <GridRow>
                <MultiChoice
                  meta={{
                    touched: confirmLandingsChangeError,
                    error: confirmLandingsChangeError,
                  }}
                  id="confirmLandingsChange"
                >
                  <fieldset>
                    <SelectRadio
                      id="landingsTypeYes"
                      checked={confirmLandingsType === 'Yes'}
                      value="Yes"
                      name="confirmLandingsChange"
                      onChange={this.onChange}
                    >
                      {t('ccLandingsTypeConfirmationOptionYesText')}
                    </SelectRadio>
                    <SelectRadio
                      id="landingsTypeNo"
                      checked={confirmLandingsType === 'No'}
                      value="No"
                      name="confirmLandingsChange"
                      onChange={this.onChange}
                    >
                      {t('ccLandingsTypeConfirmationOptionNoText')}
                    </SelectRadio>
                  </fieldset>
                </MultiChoice>
              </GridRow>
            </GridCol>
          <GridRow>
            <SecondaryButton
              type="button"
              id="cancel"
              name="cancel"
              value="cancel"
              onClick={this.onClear}
            >
               {t('commonSecondaryButtonCancelButton')}
            </SecondaryButton>
            <ContinueButton type="submit" id="continue">
              {t('commonContinueButtonContinueButtonText')}
            </ContinueButton>
          </GridRow>
          <input type="hidden" name="confirmLandingsType" value={confirmLandingsType} />
        </Form>
        <HelpLink journey={route.journey} />
      </Main>
    );
  }
}

function mapStateToProps(state) {
  return {
    errors: state.errors,
    landingsType: state.landingsType,
    changedLandingsType: state.changedLandingsType,
  };
}

function loadData(store) {
  if (Object.keys(this.queryParams).length !== 0) {
    store.dispatch(dispatchApiCallFailed(JSON.parse(this.queryParams.error)));
  }
  store.dispatch(onLoadComponentRedirect());
}

export const component = withRouter(
  connect(mapStateToProps, {
    clearLandingsType,
    confirmChangeLandingsType,
    dispatchClearErrors,
    clearChangedLandingsType
  })(withTranslation() (LandingsTypesConfirmationPage))
);

LandingsTypesConfirmationPage.propTypes = {
  errors: PropTypes.object,
};

export default {
  loadData,
  component: PageTemplateWrapper(component),
};
