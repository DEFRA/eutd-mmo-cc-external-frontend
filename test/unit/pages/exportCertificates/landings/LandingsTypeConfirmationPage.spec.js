import * as React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter, Router, Route } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { act } from 'react-dom/test-utils';
import thunk from 'redux-thunk';
import { createMemoryHistory } from 'history';
import * as Utils from '../../../../../src/client/pages/utils/errorUtils';

import { component as LandingsTypeConfirmationPage } from '../../../../../src/client/pages/exportCertificates/landings/LandingsTypeConfirmationPage';
import landingsTypeConfirmationPage from '../../../../../src/client/pages/exportCertificates/landings/LandingsTypeConfirmationPage';
import { dispatchClearErrors } from '../../../../../src/client/actions';
import { confirmChangeLandingsType, clearLandingsType, onLoadComponentRedirect, clearChangedLandingsType } from '../../../../../src/client/actions/landingsType.actions';
import { render } from '@testing-library/react';
jest.mock('../../../../../src/client/actions');
jest.mock('../../../../../src/client/actions/export-payload.actions');
jest.mock('../../../../../src/client/actions/landingsType.actions');

describe('Landings Type Confirmation Page', () => {
  let wrapper;

  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res([]);
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);

  const store = mockStore({
    errors: {},
    landingsType: { landingsEntryOption: 'manualEntry' },
    changedLandingsType: 'directLanding',
    t: jest.fn()
  });

  const props = {
    route: {
      path: '/catch-certificates/landings-type-confirmation',
      title: 'Create a UK catch certificate - GOV.UK',
      previousUri: '/catch-certificates/landings-entry',
      nextUri: '/catch-certificates/what-are-you-exporting',
      dashboardUri: 'catch-certificates',
      uploadFileUri: '/catch-certificates/upload-file',
      journey: 'catchCertificate',
      journeyText: 'catch certificate',
    },
  };

  const documentNumber = 'document123';
  const mockPreventDefault = jest.fn();
  const history = createMemoryHistory({
    initialEntries: [`/catch-certificates/${documentNumber}/landings-type-confirmation`],
    initialIndex: 0
  });
  const mockPush = jest.spyOn(history, 'push');

  wrapper = mount(
    <Provider store={store}>
      <Router history={history} >
        <Route path="/catch-certificates/:documentNumber/landings-type-confirmation">
          <LandingsTypeConfirmationPage {...props} />
        </Route>
      </Router>
    </Provider>
  );

  beforeEach(() => {
    mockPush.mockReset();
    mockPreventDefault.mockReturnValue(null);
    confirmChangeLandingsType.mockReturnValue({ type: 'EXPORT_PAYLOAD_LOADED' });
    clearLandingsType.mockReturnValue({type:'CLEAR_LANDINGS_TYPE'});
    onLoadComponentRedirect.mockReturnValue({ type: 'LANDINGS_TYPE_CHANGE_UNAUTHORISED' });
    dispatchClearErrors.mockReturnValue({ type: 'DISPATCH_CLEAR_ERRORS' });
    clearChangedLandingsType.mockReturnValue({type: 'clear_changed_landings_type'});
  });

  afterEach(() => {
    confirmChangeLandingsType.mockRestore();
    clearLandingsType.mockRestore();
    dispatchClearErrors.mockRestore();
    clearChangedLandingsType.mockRestore();
    mockPreventDefault.mockRestore();
  });

  it('should mount successfully', () => {
    expect(wrapper).toBeDefined();
  });

  it('should contain a title', () => {
    expect(wrapper.find('PageTitle').props().title).toBe('Are you sure you want to change your landings type? - Create a UK catch certificate - GOV.UK');
  });

  it('should contain a back link', () => {
    expect(wrapper.find('BackLink').exists()).toBeTruthy();
    expect(wrapper.find('BackLink').props().href).toContain('/catch-certificates/landings-entry');
  });

  it('should contain a back link that clears any errors and goes to the previous screen', () => {
    const backUri = props.route.previousUri.replace(':documentNumber', documentNumber);

    act(() => {
      wrapper.find('BackLink').prop('onClick')({ preventDefault: mockPreventDefault });
    });

    expect(mockPreventDefault).toHaveBeenCalled();
    expect(dispatchClearErrors).toHaveBeenCalled();
    expect(mockPush).toHaveBeenCalledWith(backUri);
  });

  it('should contain warning', () => {
    expect(wrapper.find('WarningText').props().className).toContain('warning-message');
    expect(wrapper.find('WarningText').text()).toContain('Switching between direct and non-direct landings types will require the re-entry of landings data.');
  });

  it('should contain radio options', () => {
    expect(wrapper.find('input#landingsTypeYes').exists()).toBeTruthy();
    expect(wrapper.find('label#label-landingsTypeYes span').text()).toBe('Yes, I want to change my landings type');

    expect(wrapper.find('input#landingsTypeNo').exists()).toBeTruthy();
    expect(wrapper.find('label#label-landingsTypeNo span').text()).toBe('No, I want to keep my current landings type');
  });

  it('should set state when Yes option selected', () => {
    act(() => {
      wrapper.find('input#landingsTypeYes').prop('onChange')({ preventDefault: () => { }, target: { value: 'Yes' } });
    });

    expect(wrapper.find('LandingsTypesConfirmationPage').state().confirmLandingsType).toBe('Yes');
  });

  it('should set state when No option selected', () => {
    act(() => {
      wrapper.find('input#landingsTypeNo').prop('onChange')({ preventDefault: () => { }, target: { value: 'No' } });
    });

    expect(wrapper.find('LandingsTypesConfirmationPage').state().confirmLandingsType).toBe('No');
  });

  it('should contain buttons', () => {
    expect(wrapper.find('button#cancel').exists()).toBeTruthy();
    expect(wrapper.find('button#continue').exists()).toBeTruthy();
  });

  it('should contain a cancel button that clears any errors and goes to the previous screen', () => {
    const backUri = props.route.previousUri.replace(':documentNumber', documentNumber);

    act(() => {
      wrapper.find('button#cancel').prop('onClick')({ preventDefault: mockPreventDefault });
    });

    expect(mockPreventDefault).toHaveBeenCalled();
    expect(dispatchClearErrors).toHaveBeenCalled();
    expect(mockPush).toHaveBeenCalledWith(backUri);
  });

  it('should contain a continue button that goes to the add products page', () => {
    wrapper.find('LandingsTypesConfirmationPage').setState({ confirmLandingsType: 'Yes' });

    act(() => {
      wrapper.find('Form').prop('onSubmit')({ preventDefault: mockPreventDefault });
    });

    expect(mockPreventDefault).toHaveBeenCalled();
    expect(confirmChangeLandingsType).toHaveBeenCalled();
    expect(confirmChangeLandingsType).toHaveBeenCalledWith(
      {
        'landingsEntryConfirmation': 'Yes',
        'landingsEntryOption': 'directLanding'
      },
      props.route.journey,
      props.route.path,
      documentNumber
    );
  });

  it('should contain a continue button that goes to the upload file page', () => {
    const store = mockStore({
      errors: {},
      landingsType: { landingsEntryOption: null },
      changedLandingsType: 'uploadEntry'
    });

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/catch-certificates/:documentNumber/landings-type-confirmation']} initialIndex={0} >
          <LandingsTypeConfirmationPage {...props} />
        </MemoryRouter>
      </Provider>
    );

    wrapper.find('LandingsTypesConfirmationPage').setState({ confirmLandingsType: 'Yes' });

    act(() => {
      wrapper.find('Form').prop('onSubmit')({ preventDefault: mockPreventDefault });
    });

    expect(mockPreventDefault).toHaveBeenCalled();
    expect(confirmChangeLandingsType).toHaveBeenCalled();
    expect(confirmChangeLandingsType).toHaveBeenCalledWith({ 'landingsEntryConfirmation': 'Yes', 'landingsEntryOption': 'uploadEntry' }, 'catchCertificate', '/catch-certificates/landings-type-confirmation', undefined);
  });

  it('should contain a help link', () => {
    expect(wrapper.find('HelpLink').exists()).toBeTruthy();
    expect(wrapper.find('HelpLink').props().journey).toBe('catchCertificate');
  });

  it('Should redirect the user to the forbidden page when there is unauthorised access to the page', () => {
    const mockPush = jest.fn();

    new LandingsTypeConfirmationPage.WrappedComponent({
      match: {
        params: {
          documentNumber: ''
        }
      },
      history: {
        push: mockPush
      },
      route: {
        path: '/catch-certificates/:documentNumber/landings-type-confirmation',
        title: 'Create a UK catch certificate - GOV.UK',
        previousUri: '/catch-certificates/:documentNumber/landings-options',
        addLandingsUri: '/catch-certificates/:documentNumber/add-landings',
        directLandingUri: '/catch-certificates/:documentNumber/direct-landing',
        journey: 'catchCertificate',
        journeyText: 'catch certificate',
      },
      landingsType: { unauthorised: true }
    }).componentDidMount();

    expect(mockPush).toHaveBeenCalledWith('/forbidden');
  });

  it('Should not redirect the user to the forbidden page when there is authorised access to the page', () => {
    const mockPush = jest.fn();

    new LandingsTypeConfirmationPage.WrappedComponent({
      match: {
        params: {
          documentNumber: ''
        }
      },
      history: {
        push: mockPush
      },
      route: {
        path: '/catch-certificates/:documentNumber/landings-type-confirmation',
        title: 'Create a UK catch certificate - GOV.UK',
        previousUri: '/catch-certificates/:documentNumber/landings-options',
        addLandingsUri: '/catch-certificates/:documentNumber/add-landings',
        directLandingUri: '/catch-certificates/:documentNumber/direct-landing',
        journey: 'catchCertificate',
        journeyText: 'catch certificate',
      },
      landingsType: { unauthorised: false }
    }).componentDidMount();

    expect(mockPush).not.toHaveBeenCalledWith('/forbidden');
  });

  it('should unmount and clear landings type', () => {
    wrapper.unmount();
    expect(clearLandingsType).toHaveBeenCalled();
    expect(clearChangedLandingsType).toHaveBeenCalled();
  });

});

describe('Snapshot describe',()=>{

  let wrapper;

  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res([]);
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);

  const store = mockStore({
    errors: {},
    landingsType: { landingsEntryOption: 'manualEntry' },
    changedLandingsType: 'directLanding',
    t: jest.fn()
  });

  const props = {
    route: {
      path: '/catch-certificates/landings-type-confirmation',
      title: 'Create a UK catch certificate - GOV.UK',
      previousUri: '/catch-certificates/landings-entry',
      nextUri: '/catch-certificates/what-are-you-exporting',
      dashboardUri: 'catch-certificates',
      uploadFileUri: '/catch-certificates/upload-file',
      journey: 'catchCertificate',
      journeyText: 'catch certificate',
    },
    t: jest.fn(),
    clearLandingsType: jest.fn()

  };

  const documentNumber = 'document123';
  const mockPreventDefault = jest.fn();
  const history = createMemoryHistory({
    initialEntries: [`/catch-certificates/${documentNumber}/landings-type-confirmation`],
    initialIndex: 0
  });
  const mockPush = jest.spyOn(history, 'push');

  wrapper = mount(
    <Provider store={store}>
      <Router history={history} >
        <Route path="/catch-certificates/:documentNumber/landings-type-confirmation">
          <LandingsTypeConfirmationPage {...props} />
        </Route>
      </Router>
    </Provider>
  );

  beforeEach(() => {
    mockPush.mockReset();
    mockPreventDefault.mockReturnValue(null);
    confirmChangeLandingsType.mockReturnValue({ type: 'EXPORT_PAYLOAD_LOADED' });
    clearLandingsType.mockReturnValue({type:'landingsType/landings_type/clear'});
    onLoadComponentRedirect.mockReturnValue({ type: 'LANDINGS_TYPE_CHANGE_UNAUTHORISED' });
    dispatchClearErrors.mockReturnValue({ type: 'DISPATCH_CLEAR_ERRORS' });
    clearChangedLandingsType.mockReturnValue({type: 'clear_changed_landings_type'});
  });

  
  it('should take a snapshot of the whole page', () => {
  const { container } = render(wrapper);
   expect(container).toMatchSnapshot();
  });
  
})

describe('Landings Type Confirmation Page with errors', () => {
  let wrapper;

  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res([]);
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);

  const store = mockStore({
    errors: {
      errors: [{
        text: 'We have an issue',
        targetName: 'confirmLandingsChange'
      }]
    },
    t: jest.fn()
  });

  const props = {
    route: {
      path: '/catch-certificates/:documentNumber/add-landings',
      title: 'Create a UK catch certificate - GOV.UK',
      previousUri: '/catch-certificates/:documentNumber/landings-options',
      addLandingsUri: '/catch-certificates/:documentNumber/add-landings',
      directLandingUri: '/catch-certificates/:documentNumber/direct-landing',
      journey: 'catchCertificate',
      journeyText: 'catch certificate',
    },
    t: jest.fn()
  };

  const mockPreventDefault = jest.fn();

  wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <LandingsTypeConfirmationPage {...props} />
      </MemoryRouter>
    </Provider>
  );

  let mockScrollToErrorIsland;

  beforeEach(() => {
    confirmChangeLandingsType.mockReturnValue({ type: 'EXPORT_PAYLOAD_LOADED' });
    mockScrollToErrorIsland = jest.spyOn(Utils, 'scrollToErrorIsland');
  });

  afterEach(() => {
    confirmChangeLandingsType.mockRestore();
    mockPreventDefault.mockRestore();
    mockScrollToErrorIsland.mockRestore();
  });

  it('should contain a title', () => {
    expect(wrapper.find('PageTitle').props().title).toBe('Error: Are you sure you want to change your landings type? - Create a UK catch certificate - GOV.UK');
  });

  it('should have an error island', () => {
    expect(wrapper.find('ErrorIsland').exists()).toBeTruthy();
    expect(wrapper.find('ErrorIsland').props().errors).toHaveLength(1);
  });

  it('should contain a continue button that goes to the next screen', () => {

    wrapper.find('LandingsTypesConfirmationPage').setState({ confirmLandingsType: 'Yes' });

    act(() => {
      wrapper.find('Form').prop('onSubmit')({ preventDefault: mockPreventDefault });
    });

    expect(mockPreventDefault).toHaveBeenCalled();
    expect(confirmChangeLandingsType).toHaveBeenCalled();
  });

  it('should contain a continue button that goes to the next screen when confirmation is no', () => {

    wrapper.find('LandingsTypesConfirmationPage').setState({ confirmLandingsType: 'No' });

    act(() => {
      wrapper.find('Form').prop('onSubmit')({ preventDefault: mockPreventDefault });
    });

    expect(mockPreventDefault).toHaveBeenCalled();
    expect(confirmChangeLandingsType).toHaveBeenCalled();
  });

  it('will call ScrollToErrorIsland if an error is thrown', () => {
    confirmChangeLandingsType.mockImplementation(() => {
      throw 'error';
    });

    act(() => {
      wrapper.find('Form').prop('onSubmit')({ preventDefault: mockPreventDefault });
    });

    expect(mockScrollToErrorIsland).toHaveBeenCalled();
  });
});

describe('Landings Type loadData', () => {

  const store = {
    dispatch: jest.fn()
  };

  const queryParams = {
    error: '{"x":5,"y":6}'
  };


  it('will call all methods needed to load the component', () => {
    landingsTypeConfirmationPage.queryParams = {};

    landingsTypeConfirmationPage.loadData(store);

    expect(onLoadComponentRedirect).toHaveBeenCalled();
  });

  it('will call all methods needed to load the component when given query params', () => {
    landingsTypeConfirmationPage.queryParams = queryParams;

    landingsTypeConfirmationPage.loadData(store);

    expect(store.dispatch).toHaveBeenCalled();
  });

});


describe('landingsTypeConfirmationPage, Page guard', () => {

  it('Should redirect the exporter to dashboard when changedLandingsType is empty', () => {
    const mockPush = jest.fn();
    const props = {
      match: {
        params: {
          documentNumber: ''
        }
      },
      history: {
        push: mockPush
      },
      route: {
        path: '/catch-certificates/:documentNumber/landings-type-confirmation',
        title: 'Create a UK catch certificate - GOV.UK',
        previousUri: '/catch-certificates/:documentNumber/landings-entry',
        nextUri: '/catch-certificates/:documentNumber/what-are-you-exporting',
        dashboardUri: 'catch-certificates',
        uploadFileUri: '/catch-certificates/:documentNumber/upload-file',
        journey: 'catchCertificate',
        journeyText: 'catch certificate',
      },
      landingsType: { landingsEntryOption: null },
      changedLandingsType: '',
      t: jest.fn()
    };

    new LandingsTypeConfirmationPage.WrappedComponent(props).componentDidMount();

    expect(mockPush).toHaveBeenCalledWith('catch-certificates');
  });
});