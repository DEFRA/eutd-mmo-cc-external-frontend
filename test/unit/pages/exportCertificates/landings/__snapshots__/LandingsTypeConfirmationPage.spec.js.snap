// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot describe should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <a
          class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
          href="/catch-certificates/landings-entry"
        >
          Back
        </a>
      </div>
      <form
        action="/orchestration/api/v1/confirm-change-landings-type?c=/catch-certificates/landings-type-confirmation"
        method="POST"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <div
            class="css-1lzr9ic-StyledRow e4u1u2g0"
          >
            <h1
              class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
            >
              <legend>
                Are you sure you want to change your landings type?
              </legend>
            </h1>
          </div>
          <div
            class="css-1lzr9ic-StyledRow e4u1u2g0"
          >
            <div
              class="warning-message e1vcssft0 css-jxvtpr-StyledWarningText-StyledHoc e4u1u2g0"
            >
              <div
                class="css-96kz36-IconImportantWrapper e4u1u2g1"
              >
                <svg
                  fill="currentColor"
                  height="100%"
                  preserveAspectRatio="xMidYMid meet"
                  style="display: block;"
                  version="1.1"
                  viewBox="0 0 35.000000 35.000000"
                  width="100%"
                >
                  <title>
                    icon important
                  </title>
                  <g
                    transform="translate(0.000000,35.000000) scale(0.100000,-0.100000)"
                  >
                    <path
                      d="M100 332 c-87 -48 -125 -155 -82 -232 48 -87 155 -125 232 -82 87 48
125 155 82 232 -48 87 -155 125 -232 82z m100 -122 c0 -53 -2 -60 -20 -60 -18
0 -20 7 -20 60 0 53 2 60 20 60 18 0 20 -7 20 -60z m0 -111 c0 -12 -7 -19 -20
-19 -19 0 -28 28 -14 43 11 11 34 -5 34 -24z"
                    />
                  </g>
                </svg>
              </div>
              <strong
                class="css-jegmsb-WarningTextWrapper e4u1u2g2"
              >
                Switching between direct and non-direct landings types will require the re-entry of landings data.
              </strong>
            </div>
          </div>
          <div
            class="css-1lzr9ic-StyledRow e4u1u2g0"
          >
            <div
              class="e1vcssft0 css-1phgi8s-StyledFieldset-StyledHoc e4u1u2g0"
              id="confirmLandingsChange"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              />
              <fieldset>
                <label
                  class="css-sdrcg6-StyledHoc"
                  for="landingsTypeYes"
                  id="label-landingsTypeYes"
                >
                  <input
                    class="css-1xt3i3l"
                    id="landingsTypeYes"
                    name="confirmLandingsChange"
                    type="radio"
                    value="Yes"
                  />
                  <span
                    class="css-1uwxoxa"
                  >
                    Yes, I want to change my landings type
                  </span>
                </label>
                <label
                  class="css-sdrcg6-StyledHoc"
                  for="landingsTypeNo"
                  id="label-landingsTypeNo"
                >
                  <input
                    class="css-1xt3i3l"
                    id="landingsTypeNo"
                    name="confirmLandingsChange"
                    type="radio"
                    value="No"
                  />
                  <span
                    class="css-1uwxoxa"
                  >
                    No, I want to keep my current landings type
                  </span>
                </label>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            id="cancel"
            name="cancel"
            type="button"
            value="cancel"
          >
            Cancel
          </button>
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
            type="submit"
          >
            Continue
          </button>
        </div>
        <input
          name="confirmLandingsType"
          type="hidden"
          value=""
        />
      </form>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
